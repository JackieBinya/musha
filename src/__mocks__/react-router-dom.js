module.exports = {
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: jest.fn(),
  useLocation: jest.fn(),
};
