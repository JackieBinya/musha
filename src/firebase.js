import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/storage';

const firebaseConfig = firebase.initializeApp({
  apiKey: 'AIzaSyDD14Pef51sWG0czExRugPOLJO-Ne44bNI',
  authDomain: 'musha-000.firebaseapp.com',
  databaseURL: 'https://musha-000.firebaseio.com',
  projectId: 'musha-000',
  storageBucket: 'musha-000.appspot.com',
  messagingSenderId: '111803589651',
  appId: '1:111803589651:web:b90b953a233696281b4be6',
});

const storage = firebase.storage();

export { storage, firebaseConfig as firebase };
