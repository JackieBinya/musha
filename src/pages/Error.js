import React from 'react';

export const Error = () => <h2>Error</h2>;