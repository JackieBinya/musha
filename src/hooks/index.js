import React, { useState, useEffect } from 'react';
import { firebase } from '../firebase';

export const useProperties = () => {
  const [properties, setProperties] = useState([]);

  useEffect(() => {
    firebase
      .firestore()
      .collection('properties')
      .get()
      .then((data) => {
        const allProperties = data.docs.map((doc) => ({
          ...doc.data(),
          id: doc.id,
        }));

        setProperties([...allProperties]);
      });
  }, []);

  return { properties, setProperties };
};

export const useProperty = (id) => {
  const [property, setProperty] = useState('');

  useEffect(() => {
    firebase
      .firestore()
      .collection('properties')
      .doc(id)
      .get()
      .then((doc) => {
        if (doc.exists) {
          const newProperty = doc.data();
          setProperty(newProperty);
        }
      })
      .catch((error) => console.log(error));
  }, []);

  return { property };
};

export const usePropertiesByUserID = (userId) => {
  const [userProperties, setUserProperties] = useState([]);
  console.log(userProperties);

  useEffect(() => {
    firebase
      .firestore()
      .collection('properties')
      .where('ownerID', '==', userId)
      .get()
      .then((data) => {
        const newProperties = data.docs.map((doc) => ({
          ...doc.data(),
          id: doc.id,
        }));

        setUserProperties([...newProperties]);
      })
      .catch((error) => console.log(error));
  }, []);

  return { userProperties };
};
