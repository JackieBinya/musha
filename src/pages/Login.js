import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { SignUp } from './SignUp';
import { FormMessage } from '../components/FormMessage';
import { PasswordInput } from '../components/PasswordInput';
import { EmailInput } from '../components/EmailInput';
import { firebase } from '../firebase';

export const Login = ({ history }) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [message, setMessage] = useState('');
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const handleLogin = async (e) => {
    e.preventDefault();

    try {
      await firebase.auth().signInWithEmailAndPassword(email, password);
      setIsAuthenticated(true);
      setMessage('Congrats 🥳, you are successfully logged in.');
      setTimeout(() => history.push('/my-properties'), 5000);
    } catch (error) {
      setIsAuthenticated(false);

      if (error.code === ' auth/invalid-email') {
        setMessage('🤖 email address is not valid');
      }
      if (error.code === 'auth/user-disabled') {
        setMessage('🤖 your account has been suspended');
      }
      if (error.code === 'auth/user-not-found') {
        setMessage('🤖 email does no exist!');
      }
      if (error.code === ' auth/wrong-password') {
        setMessage('🤖 incorrect password');
      }
    }
  };
  return (
    <>
      <Link to="/signup">Sign Up</Link>
      <h1>Login</h1>
      <form className="auth-form" onSubmit={handleLogin}>
        <FormMessage
          isAuthenticated={isAuthenticated}
          message={message}
          setMessage={setMessage}
        />
        <EmailInput email={email} setEmail={setEmail} />
        <PasswordInput setPassword={setPassword} password={password} />
        <button type="submit">Continue</button>
      </form>
      <Link to="/password-reset">Have you forgotten your password?</Link>
    </>
  );
};
