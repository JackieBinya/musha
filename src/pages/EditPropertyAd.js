import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import { firebase } from '../firebase';
import { yupResolver } from '@hookform/resolvers';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';

import { useProperty } from '../hooks';
import { FirstSection } from '../components/FirstSection';
import { LocationSection } from '../components/LocationSection';
import { Loader } from '../components/Loader';
import { PropertyDetailsSection } from '../components/PropertyDetailsSection';
import { PROPERTY_SCHEMA } from '../constants/validations';

export const EditPropertyAds = ({ history }) => {
  const { propertyId } = useParams();
  const { property } = useProperty(propertyId);

  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(PROPERTY_SCHEMA),
  });

  const [isTitleUpdated, setIsTitleUpdated] = useState(false);
  const [isMobileNumberUpdated, setIsMobileNumberUpdated] = useState(false);
  const [isAvailableToUpdated, setIsAvailableToUpdated] = useState(false);
  const [isCityUpdated, setIsCityUpdated] = useState(false);
  const [isLocationUpdated, setIsLocationUpdated] = useState(false);
  const [isNumberOfBathroomsUpdated, setIsNumberOfBathroomsUpdated] = useState(
    false
  );
  const [isNumberOfBedroomsUpdated, setIsNumberOfBedroomsUpdated] = useState(
    false
  );
  const [isDescriptionUpdated, setIsDescriptionUpdated] = useState(false);

  const [isModalVisible, setIsModalVisible] = useState(false);

  const onSubmit = (
    {
      city,
      location,
      availableTo,
      numberOfBathrooms,
      numberOfBedrooms,
      description,
      mobileNumber,
      title,
    },
    e
  ) => {
    const propertyRef = firebase
      .firestore()
      .collection('properties')
      .doc(propertyId);

    if (property.title !== title) {
      propertyRef
        .update({
          title,
        })
        .then(() => setIsTitleUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.availableTo !== availableTo) {
      propertyRef
        .update({
          availableTo,
        })
        .then(() => setIsAvailableToUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.mobileNumber !== mobileNumber) {
      propertyRef
        .update({
          mobileNumber,
        })
        .then(() => setIsMobileNumberUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.city !== city) {
      propertyRef
        .update({
          city,
        })
        .then(() => setIsCityUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.location !== location) {
      propertyRef
        .update({
          location,
        })
        .then(() => setIsLocationUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.description !== description) {
      propertyRef
        .update({
          description,
        })
        .then(() => setIsDescriptionUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.numberOfBedrooms !== numberOfBedrooms) {
      propertyRef
        .update({
          numberOfBedrooms,
        })
        .then(() => setIsNumberOfBedroomsUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.numberOfBathrooms !== numberOfBathrooms) {
      propertyRef
        .update({
          numberOfBathrooms,
        })
        .then(() => setIsNumberOfBathroomsUpdated(true))
        .catch((error) => console.log(error));
    }
  };

  useEffect(() => {
    if (
      isTitleUpdated ||
      isMobileNumberUpdated ||
      isAvailableToUpdated ||
      isCityUpdated ||
      isLocationUpdated ||
      isNumberOfBathroomsUpdated ||
      isNumberOfBedroomsUpdated ||
      isDescriptionUpdated
    ) {
      setIsModalVisible(true);

      setIsTitleUpdated(false);
      setIsAvailableToUpdated(false);
      setIsMobileNumberUpdated(false);
      setIsCityUpdated(false);
      setIsLocationUpdated(false);
      setIsNumberOfBathroomsUpdated(false);
      setIsNumberOfBedroomsUpdated(false);
      setIsDescriptionUpdated(false);
    }
  }, [
    isTitleUpdated,
    isAvailableToUpdated,
    isCityUpdated,
    isLocationUpdated,
    isMobileNumberUpdated,
    isDescriptionUpdated,
    isNumberOfBathroomsUpdated,
    isNumberOfBedroomsUpdated,
  ]);

  return (
    <div className="container">
      <div style={{ marginBottom: '2em' }}>
        {property && isModalVisible && (
          <div className="form-modal-container">
            <p
              className={isModalVisible ? 'show' : 'hide'}
            >{`Property ID:${propertyId} is successfully updated`}</p>
            <button
              data-testid="go-to-my-properties-action"
              type="button"
              onClick={() => history.goBack()}
            >
              {' '}
              Go to my properties
            </button>
          </div>
        )}
        <div className="property-form-container">
          {!property ? (
            <Loader />
          ) : (
            <div
              data-testid="edit-form"
              className={isModalVisible ? 'hide' : 'show'}
            >
              <button
                data-testid="back-button"
                type="button"
                onClick={() => history.goBack()}
              >
                <FontAwesomeIcon
                  icon={faChevronLeft}
                  className="property-icons-svg"
                />
                Back
              </button>

              <h1>Edit property ad</h1>

              <form onSubmit={handleSubmit(onSubmit)} className="form-primary">
                <FirstSection
                  defaultMobileNumber={property.mobileNumber}
                  defaultTitle={property.title}
                  defaultAvailableTo={property.availableTo}
                  register={register}
                  errors={errors}
                />

                <LocationSection
                  defaultCity={property.city}
                  defaultLocation={property.location}
                  register={register}
                  errors={errors}
                />

                {/* <UploadImagesSection
                  defaultImageUrls={property.imageUrls}
                  setImageUrls={setImageUrls}
                /> */}

                <PropertyDetailsSection
                  defaultNumberOfBathrooms={property.numberOfBathrooms}
                  defaultNumberOfBedrooms={property.numberOfBedrooms}
                  defaultDescription={property.description}
                  register={register}
                  errors={errors}
                />

                <button data-testid="submit-action" type="submit">
                  Save
                </button>
              </form>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};
