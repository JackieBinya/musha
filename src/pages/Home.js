import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Fuse from 'fuse.js';
import { useProperties } from '../hooks';
import { ShortProperty } from '../components/ShortProperty';
import { Hero } from '../components/Hero';
import { Loader } from '../components/Loader';
import { PropertyIconsHome } from '../components/PropertyIconsHome';

export const Home = () => {
  const { properties } = useProperties();
  const [query, setQuery] = useState('');

  const fuse = new Fuse(properties, {
    keys: ['location', 'city'],
    includeScore: true,
  });

  const results = fuse.search(query);
  const propertiesResults = query
    ? results.map((character) => character.item)
    : properties;

  return (
    <>
      <Hero query={query} setQuery={setQuery} />
      <main>
        <div className="main container">
          {propertiesResults.length === 0 ? (
            <Loader />
          ) : propertiesResults ? (
            propertiesResults.map(
              ({
                id,
                location,
                imageUrls,
                numberOfBathrooms,
                numberOfBedrooms,
                city,
                title,
                description,
              }) => (
                <li key={id} className="main-list-item">
                  <Link to={`/property/${id}`} className="main-link">
                    <ShortProperty
                      imageUrls={imageUrls}
                      title={title}
                      location={location}
                      description={description}
                    >
                      <PropertyIconsHome
                        city={city}
                        title={title}
                        location={location}
                        numberOfBathrooms={numberOfBathrooms}
                        numberOfBedrooms={numberOfBedrooms}
                      />
                    </ShortProperty>
                  </Link>
                </li>
              )
            )
          ) : (
            ''
          )}
        </div>
      </main>
    </>
  );
};
