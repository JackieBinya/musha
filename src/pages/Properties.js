import React, { useState, useContext } from 'react';
import { firebase } from '../firebase';
import { Link, Route, useRouteMatch } from 'react-router-dom';
import { PostPropertyAd } from './PostPropertyAd';
import { usePropertiesByUserID } from '../hooks';
import { AuthContext } from '../context/auth-context';
import { Loader } from '../components/Loader';
import { ShortProperty } from '../components/ShortProperty';
import { PropertyIcons } from '../components/PropertyIcons';

/* 
Authenticated users access a form they can use to list a new property
Users get their listed properties by id'
Users can update their properties
*/

export const Properties = ({ history }) => {
  const { path } = useRouteMatch();
  const { currentUser } = useContext(AuthContext);
  const { userProperties } = usePropertiesByUserID(currentUser.uid);

  const handleLogOut = () => {
    firebase.auth().signOut();
    history.push('/');
  };

  return (
    <div className="container">
      <h1>Properties</h1>
      <button type="button" onClick={handleLogOut}>
        Logout
      </button>

      <Link to={`${path}/post`}>Post a property</Link>

      {!userProperties ? (
        <Loader />
      ) : (
        userProperties.map(
          ({
            city,
            location,
            imageUrls,
            numberOfBathrooms,
            numberOfBedrooms,
            title,
            description,
            id,
          }) => (
            <div key={id} className="my-properties">
              <h3>{title ? title : `New property in ${location}`}</h3>
              <ShortProperty imageUrls={imageUrls}>
                <h3 className="property-id">Property ID: {id}</h3>
                <p>{description}</p>
                <PropertyIcons
                  id={id}
                  city={city}
                  location={location}
                  numberOfBathrooms={numberOfBathrooms}
                  numberOfBedrooms={numberOfBedrooms}
                  user={currentUser}
                  path={path}
                />
              </ShortProperty>
            </div>
          )
        )
      )}
    </div>
  );
};
