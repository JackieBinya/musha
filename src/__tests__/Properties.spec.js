import React from 'react';
import { firebase } from '../firebase';
import { usePropertiesByUserID } from '../hooks';
import { useRouteMatch, useParams } from 'react-router-dom';
import {
  render,
  cleanup,
  fireEvent,
  act,
  waitFor,
} from '@testing-library/react';
import { Properties } from '../pages/Properties';

import { toHaveClass } from '@testing-library/jest-dom/matchers';

expect.extend({ toHaveClass });

jest.mock('react', () => {
  const ActualReact = require.requireActual('react');
  return {
    ...ActualReact,
    useContext: (data) => ({
      currentUser: {
        uid: '123456',
        email: 'foo@bar.com',
      },
    }),
  };
});

jest.mock('../hooks', () => ({
  usePropertiesByUserID: jest.fn(),
}));

describe('<Properties/>', () => {
  useRouteMatch.mockReturnValue({ path: '/my-properties' });

  it('The loader component is rendered when making API calls to fetch resource.', () => {
    usePropertiesByUserID.mockReturnValue({ userProperties: null });

    const { queryByText } = render(<Properties />);

    expect(queryByText('Properties')).toBeTruthy();
    expect(queryByText('Loading...')).toBeTruthy();
  });

  it('The user receives an alert notification if there are no properties.', () => {
    usePropertiesByUserID.mockReturnValue({ userProperties: [] });

    const { queryByText, debug } = render(<Properties />);

    expect(queryByText('Properties')).toBeTruthy();
    expect(
      queryByText(
        'Ooops! No properties found, start posting property ads for free.'
      )
    ).toBeTruthy();
  });

  it('If properties are found they are rendered', () => {
    usePropertiesByUserID.mockReturnValue({
      userProperties: [
        {
          id: '1',
          city: 'Test City',
          location: 'Test Location',
          availableTo: 'rent',
          numberOfBathrooms: '2',
          numberOfBedrooms: '3',
          description: 'Test Description',
          mobileNumber: '123456',
          title: 'Test Title',
          imageUrls: ['https://test.image.com'],
        },
      ],
    });

    const { asFragment } = render(<Properties />);

    expect(asFragment()).toMatchSnapshot();
  });
});
