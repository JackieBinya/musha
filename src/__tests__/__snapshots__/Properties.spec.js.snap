// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Properties/> If properties are found they are rendered 1`] = `
<DocumentFragment>
  <div
    class="container"
    style="margin-bottom: 6em;"
  >
    <h1>
      Properties
    </h1>
    <div>
      <div>
        <div
          class="my-properties"
        >
          <h2>
            Test Title
          </h2>
          <section
            class="short-property default-pointer"
          >
            <div
              class="short-property-image-container"
            >
              <img
                alt="property"
                class="short-property-image"
              />
              <div
                class="icon-container"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-camera fa-w-16 "
                  data-icon="camera"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M512 144v288c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V144c0-26.5 21.5-48 48-48h88l12.3-32.9c7-18.7 24.9-31.1 44.9-31.1h125.5c20 0 37.9 12.4 44.9 31.1L376 96h88c26.5 0 48 21.5 48 48zM376 288c0-66.2-53.8-120-120-120s-120 53.8-120 120 53.8 120 120 120 120-53.8 120-120zm-32 0c0 48.5-39.5 88-88 88s-88-39.5-88-88 39.5-88 88-88 88 39.5 88 88z"
                    fill="currentColor"
                  />
                </svg>
                <span>
                  1
                </span>
              </div>
            </div>
            <div
              class="short-property-right"
            >
              <div
                class="short-property-title hide"
              >
                 New property in undefined
              </div>
              <div
                class="hide"
              >
                <p
                  class="short-property-description"
                />
              </div>
              <h3
                class="property-id"
              >
                Property ID: 1
              </h3>
              <div
                style="padding: 1em;"
              >
                <div
                  class="display-text-group"
                >
                  <span
                    class="displayed-text"
                  >
                    Test Description 
                  </span>
                </div>
              </div>
              <section
                class="property-icons"
              >
                <div
                  class="details-icons"
                >
                  <div
                    class="property-icons-container"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-bed fa-w-20 property-icons-svg"
                      data-icon="bed"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 640 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M176 256c44.11 0 80-35.89 80-80s-35.89-80-80-80-80 35.89-80 80 35.89 80 80 80zm352-128H304c-8.84 0-16 7.16-16 16v144H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v352c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-48h512v48c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16V240c0-61.86-50.14-112-112-112z"
                        fill="currentColor"
                      />
                    </svg>
                    <span
                      class="property-icons-text"
                    >
                      3
                    </span>
                  </div>
                  <div
                    class="property-icons-container"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-bath fa-w-16 property-icons-svg"
                      data-icon="bath"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M32,384a95.4,95.4,0,0,0,32,71.09V496a16,16,0,0,0,16,16h32a16,16,0,0,0,16-16V480H384v16a16,16,0,0,0,16,16h32a16,16,0,0,0,16-16V455.09A95.4,95.4,0,0,0,480,384V336H32ZM496,256H80V69.25a21.26,21.26,0,0,1,36.28-15l19.27,19.26c-13.13,29.88-7.61,59.11,8.62,79.73l-.17.17A16,16,0,0,0,144,176l11.31,11.31a16,16,0,0,0,22.63,0L283.31,81.94a16,16,0,0,0,0-22.63L272,48a16,16,0,0,0-22.62,0l-.17.17c-20.62-16.23-49.83-21.75-79.73-8.62L150.22,20.28A69.25,69.25,0,0,0,32,69.25V256H16A16,16,0,0,0,0,272v16a16,16,0,0,0,16,16H496a16,16,0,0,0,16-16V272A16,16,0,0,0,496,256Z"
                        fill="currentColor"
                      />
                    </svg>
                    <span
                      class="property-icons-text"
                    >
                      2
                    </span>
                  </div>
                  <div
                    class="property-icons-container"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-map-marker fa-w-12 property-icons-svg"
                      data-icon="map-marker"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 384 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M172.268 501.67C26.97 291.031 0 269.413 0 192 0 85.961 85.961 0 192 0s192 85.961 192 192c0 77.413-26.97 99.031-172.268 309.67-9.535 13.774-29.93 13.773-39.464 0z"
                        fill="currentColor"
                      />
                    </svg>
                    <span
                      class="property-icons-text"
                    >
                      Test Location
                    </span>
                  </div>
                </div>
              </section>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
